### PR TITLE
`PaginationResult` should correctly implement `IteratorAggregate::getIterator()` and `Countable::count()`

### LINKED ISSUE #1: `PaginationResult` should implement `Countable`

### LINKED ISSUE #14: `PaginationResult` should implement `Countable`

### LINKED ISSUE #24: `PaginationResult` should implement `Countable`

### LINKED ISSUE #25: Bug: `PaginationResult::$records` should not be wrapped with `new ArrayIterator()` if it already implements `Traversable`
Related: [[Proposal] Blade @foreach should accept not only array and Countable but also Traversable · Issue #892 · laravel/internals](https://github.com/laravel/internals/issues/892)

### PR DESCRIPTION
**Pull request by mpyw**
This closes #24, closes #25, closes lampager/lampager-laravel#14 and closes lampager/lampager-idiorm#1

### CODE REVIEW

File: src/PaginationResult.php [hunk 1]
```diff
@@ -50,10 +50,20 @@ public function __construct($rows, array $meta)
...
+    }
+
+    /**
+     * Count records.
+     *
+     * @return int
+     */
+    public function count()
+    {
+        return $this->records instanceof \Countable || is_array($this->records) ? count($this->records) : 0;
```
Comment (chitoku-k) [inline src/PaginationResult.php]: Why does it return `0` when value is unavailable? `count` basically returns 1 when it is not countable.

Comment (mpyw) [reply src/PaginationResult.php]: Is this behavior still available in next generation PHP?

Comment (mpyw) [reply src/PaginationResult.php]: [PHP: rfc:counting_non_countables](https://wiki.php.net/rfc/counting_non_countables)

Comment (mpyw) [reply src/PaginationResult.php]: Should we return `1` triggering `E_USER_WARNING`?

Comment (chitoku-k) [reply src/PaginationResult.php]: I think so.

File: src/PaginationResult.php [hunk 2]
```diff
@@ -50,10 +50,30 @@ public function __construct($rows, array $meta)
...
+    public function count()
+    {
+        // @codeCoverageIgnoreStart
+        if (
+            !$this->records instanceof \Countable
+            && !is_array($this->records)
+            && version_compare(PHP_VERSION, '7.2.0', '>=')
+        ) {
+            // PHP: rfc:counting_non_countables https://wiki.php.net/rfc/counting_non_countables
+            trigger_error('count(): Parameter must be an array or an object that implements Countable', E_USER_WARNING);
```
Comment (chitoku-k) [inline src/PaginationResult.php]: This may be emitted by PHP >= 7.2.0, which means we should return here.

File: src/PaginationResult.php [hunk 3]
```diff
@@ -64,6 +64,16 @@ public function getIterator()
...
     {
-        return $this->records instanceof \Countable || is_array($this->records) ? count($this->records) : 0;
+        // @codeCoverageIgnoreStart
+        if (
+            !$this->records instanceof \Countable
+            && !is_array($this->records)
+            && version_compare(PHP_VERSION, '7.2.0', '>=')
+        ) {
+            // PHP: rfc:counting_non_countables https://wiki.php.net/rfc/counting_non_countables
+            trigger_error('count(): Parameter must be an array or an object that implements Countable', E_USER_WARNING);
```
Comment (mpyw) [inline src/PaginationResult.php]: This message is copied from here: 

[Online PHP editor | output for jY1aj](https://3v4l.org/jY1aj)

File: src/PaginationResult.php (hunk 2 above)
Comment (mpyw) [reply src/PaginationResult.php]: ?

Comment (mpyw) [reply src/PaginationResult.php]: I get it, we should just call `count()`.

File: src/PaginationResult.php [hunk 4]
```diff
@@ -61,19 +61,10 @@ public function getIterator()
...
-        if (
-            !$this->records instanceof \Countable
-            && !is_array($this->records)
-            && version_compare(PHP_VERSION, '7.2.0', '>=')
-        ) {
-            // PHP: rfc:counting_non_countables https://wiki.php.net/rfc/counting_non_countables
-            trigger_error('count(): Parameter must be an array or an object that implements Countable', E_USER_WARNING);
-        }
-        // @codeCoverageIgnoreEnd
         return count($this->records);
```
Comment (mpyw) [inline src/PaginationResult.php]: Just calling `count($this->records)` will handle argument as expected.

Comment (chitoku-k) [reply src/PaginationResult.php]: LGTM👍

**Pull request merged**